rectools/visuals/metrics_app.py

```diff
@@ -85,125 +97,264 @@ def construct(
+
         with fig_widget.batch_update():
-            for model in self.model_names:
-                df = data[data[Columns.Model] == model]
-                hover = f"<b>{model}</b><br>{metric_x.value}: %{{x}}<br>{metric_y.value}: %{{y}}<extra></extra>"
-                if model in existing_traces:  # pragma: no cover
-                    trace = existing_traces[model]
-                    trace.x = df[metric_x.value]
-                    trace.y = df[metric_y.value]
-                    trace.hovertemplate = hover
-                else:
-                    fig_widget.add_scatter(
-                        x=df[metric_x.value],
-                        y=df[metric_y.value],
-                        mode="markers",
-                        name=model,
-                        hovertemplate=hover,
-                        showlegend=self.show_legend,
-                    )
-            fig_widget.layout.xaxis.title = metric_x.value
-            fig_widget.layout.yaxis.title = metric_y.value
-        self.fig = go.Figure(data=fig_widget.data, layout=fig_widget.layout)
-        self.fig.layout.margin = None
+            for idx, trace in enumerate(self.fig.data):
+                if self._split_to_meta_and_model(trace.name)[0] == "nan":
+                    trace.marker.color = NAN_COLOR
+                fig_widget.data[idx].x = trace.x
+                fig_widget.data[idx].y = trace.y
+                fig_widget.data[idx].marker.color = trace.marker.color
+                fig_widget.data[idx].marker.symbol = trace.marker.symbol
+                fig_widget.data[idx].text = trace.text
+                fig_widget.data[idx].name = trace.name
+                fig_widget.data[idx].legendgroup = trace.legendgroup
+                fig_widget.data[idx].hoverinfo = trace.hoverinfo
+                fig_widget.data[idx].hovertemplate = trace.hovertemplate
+
+            fig_widget.layout = self.fig.layout
+        self.fig.layout.margin = None  # keep separate chart non-truncated
 
     def _update_fold_visibility(self, use_avg: widgets.Checkbox, fold_i: widgets.Dropdown) -> None:
         fold_i.layout.visibility = "hidden" if use_avg.value else "visible"
 
+    def _update_meta_visibility(self, use_meta: widgets.Checkbox, meta_feature: widgets.Dropdown) -> None:
+        meta_feature.layout.visibility = "hidden" if not use_meta.value else "visible"
+
+    def _create_chart_data(
+        self, use_avg: widgets.Checkbox, fold_i: widgets.Dropdown
+    ) -> pd.DataFrame:  # pragma: no cover
+        if use_avg.value or fold_i.value is None:
+            return self._make_chart_data_avg()
+        return self._make_chart_data_fold(fold_i.value)
+
     def display(self) -> None:
         """Display MetricsApp widget"""
         metric_x = widgets.Dropdown(description="Metric X:", value=self.metric_names[0], options=self.metric_names)
         metric_y = widgets.Dropdown(description="Metric Y:", value=self.metric_names[-1], options=self.metric_names)
```

Comment:
The thing is that this widget can work even with one metric - we can visualize the distance between points along the selected metric axis.

I think this case is exotic, but not useless.

Reply:
It may fail with IndexError, but there is no sense to build this plot if there is only 1 metric, right? We can add a check that we have at least 2 metrics.

Or just take smth like `self.metric_names[min(1, len(metric_names) - 1)]`.

My point was it's a bit weird to take the last value by default

Reply:
Yes, I see your point. Updated tests, now we need at least two metric columns. Set index as [0] and [1] accordingly.

Reply:
I just noticed that I wrote my comment after yours, didn't see it before. But nevertheless, it's still relevant.  
Let's handle case with 1 metric somehow, but let's take first 2 metrics by default, not first and last

Reply:
Ok, set minimal amount of metric columns back to one and handle index issue as you suggested above.